accepted/wasm-gc.md

```diff
@@ -0,0 +1,1319 @@
+
+The `GcCompiler` trait is used when compiling Wasm code. We will add a
+`gc_compiler` method to `wasmtime_environ::FuncEnvironment` to get the
+`GcCompiler` for the current function compilation:
+
+```rust
+trait FuncEnvironment {
+    // <existing methods>
+
+    fn gc_compiler(&mut self) -> &mut dyn GcCompiler;
+}
+```
+
+Fundamentally, the `GcCompiler` trait allows the collector to do three things:
+
+1. Define the layout for GC objects.
+
+2. Emit read and write barriers for struct field and array element accesses in
+   Wasm code, if necessary.
+
+3. Emit code sequences to get the type id of an object. We will use this when
+   implementing dynamic checks for downcasting references.
+
+I've sketched the trait and related types below, but I fully expect this to
+change during implementation when faced with reality. I do not expect those
+three bits of functionality to fundamentally change, however, even if we
+completely change or remove `GcStructLayout`, for example.
+
+```rust
+/// The layout of a GC object.
+pub enum GcObjectLayout {
+    Struct(GcStructLayout),
+    Array(GcArrayLayout),
+}
+
+/// The layout for a GC-managed struct type.
+pub struct GcStructLayout {
+    /// The size of this struct.
+    pub size: u32,
+    /// The alignment of this struct.
+    pub align: u32,
+    /// The fields of this struct. Each entry is the field's offset in the
+    /// struct and its type.
+    pub fields: Vec<(u32, wasm_types::WasmType)>,
+}
+
+/// The layout for a GC-managed array type.
+pub struct GcArrayLayout {
+    /// The alignment of this array and its header. Must be at least 4 bytes
+    /// so that the `len: u32` field can be aligned, but may need to be
+    /// larger if the GC requires that or if the element type requires it.
+    pub align: u32,
+    /// The size of the array header. Must be at least 4 bytes, as it must
+    /// contain a `len: u32` field indicating how many array elements follow
+    /// inline after this object.
+    pub header_size: u32,
+    /// The offset of the `len: u32` field. Must be aligned to 4 bytes.
+    pub len_field: u32,
+    /// The type of this array's elements.
+    pub element_type: wasm_types::WasmType,
+}
+
+/// A trait for the collector to hook into compilation, emitting any
+/// read/write barriers it requires.
+pub trait GcCompiler {
+    /// Define the object layouts for the types in a recursion group.
+    ///
+    /// Because the collector determines object layout, it can insert any
+    /// object headers it requires.
+    fn rec_group_layouts(&mut self, sub_types: &[wasm_types::SubType]) -> Vec<GcObjectLayout>;
+
+    /// Return the CLIF type for GC references with this collector. Must be
+    /// either `r32` or `r64`.
+    fn reference_type(&mut self, isa: &dyn TargetIsa) -> ir::types::Type;
+
+    /// The function builder type for this compiler (that is,
+    /// `cranelift_frontend::FunctionBuilder` or the Winch equivalent).
+    type FunctionBuilder;
```

Comment:
FWIW this'll be funky because this make `GcCompiler` not object safe

Reply:
I think not being object safe will actually be okay since we monomorphize based on func env anyways. Will update the part of the RFC that assumes object safety.